stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -2039,40 +2057,75 @@ swift_func_getParameterTypeInfo(
   if (!(parameterList && parameterList->getNumChildren() == typesLength))
     return -2;
 
+  SubstGenericParametersFromMetadata substFn(genericEnv, genericArguments);
+
   DecodedMetadataBuilder builder(
       demangler,
       /*substGenericParam=*/
-      [](unsigned, unsigned) { return nullptr; },
+      [&substFn](unsigned depth, unsigned index) {
+        return substFn.getMetadata(depth, index);
+      },
       /*SubstDependentWitnessTableFn=*/
-      [](const Metadata *, unsigned) { return nullptr; });
+      [&substFn](const Metadata *type, unsigned index) {
+        return substFn.getWitnessTable(type, index);
+      });
   TypeDecoder<DecodedMetadataBuilder> decoder(builder);
 
-  auto typeIdx = 0;
   // for each parameter (TupleElement), store it into the provided buffer
-  for (auto *parameter : *parameterList) {
+  for (unsigned index = 0; index != typesLength; ++index) {
+    auto *parameter = parameterList->getChild(index);
+
     if (parameter->getKind() == Node::Kind::TupleElement) {
       assert(parameter->getNumChildren() == 1);
       parameter = parameter->getFirstChild();
     }
 
     assert(parameter->getKind() == Node::Kind::Type);
 
-    auto builtTypeOrError = decoder.decodeMangledType(parameter);
-    if (builtTypeOrError.isError()) {
-      auto err = builtTypeOrError.getError();
-      char *errStr = err->copyErrorString();
-      err->freeErrorString(errStr);
-      typeIdx += 1;
-      continue;
-    }
+    auto type = decodeType(decoder, parameter);
+    if (!type)
+      return -3; // Failed to decode a type.
 
-    types[typeIdx] = builtTypeOrError.getType();
-    ++typeIdx;
+    types[index] = type;
   } // end foreach parameter
 
   return typesLength;
 }
 
+SWIFT_CC(swift)
+SWIFT_RUNTIME_STDLIB_SPI
+BufferAndSize
+swift_distributed_getWitnessTables(GenericEnvironmentDescriptor *genericEnv,
+                                   const void *const *genericArguments) {
+  assert(genericEnv);
+  assert(genericArguments);
+
+  llvm::SmallVector<const void *, 4> witnessTables;
+  SubstGenericParametersFromMetadata substFn(genericEnv, genericArguments);
+
+  auto error = _checkGenericRequirements(
+      genericEnv->getGenericRequirements(), witnessTables,
+      [&substFn](unsigned depth, unsigned index) {
+        return substFn.getMetadata(depth, index);
+      },
+      [&substFn](const Metadata *type, unsigned index) {
+        return substFn.getWitnessTable(type, index);
+      });
+
+  if (error) {
+    return {/*ptr=*/nullptr, -1};
+  }
+
+  if (witnessTables.empty())
+    return {/*ptr=*/nullptr, 0};
+
+  void **tables = (void **)malloc(witnessTables.size() * sizeof(void *));
+  for (unsigned i = 0, n = witnessTables.size(); i != n; ++i)
+    tables[i] = const_cast<void *>(witnessTables[i]);
+
+  return {tables, static_cast<intptr_t>(witnessTables.size())};
```

Comment:
cool, so no back and forth for the alloc and we just return :)

Reply:
Yeah, I wasn't sure whether that was correct but apparently it is.